File: libraries/RK4/RK4.cpp
#include "RK4.h"

//Constructor
RK4::RK4() {

};

void RK4::init(int n,double dt) {
  //Initialize State Vector
  TIMESTEP = dt;
  NUMVARS = n;
  State.zeros(NUMVARS,1,"State");
  StateDel.zeros(NUMVARS,1,"StateDel");
  k.zeros(NUMVARS,1,"k");
  k1.zeros(NUMVARS,1,"k1");
  k2.zeros(NUMVARS,1,"k2");
  k3.zeros(NUMVARS,1,"k3");
  k4.zeros(NUMVARS,1,"k4");
  phi.zeros(NUMVARS,1,"phi");
}


void RK4::set_ICs(MATLAB x0) {
  State.overwrite(x0);
  StateDel.overwrite(x0);
}

void RK4::integrate(int i) {
  //First call
  switch (i) {
  case (1):
    //Derivatives(State,k1);
    k1.overwrite(k);
    StateDel.overwrite(State); //StateDel = State
    StateDel.plus_mult_eq(k1,TIMESTEP/2); //StateDel = StateDel + k1*TIMESTEP/2
    break;
  case (2):
    //Derivatives(StateDel,k2);
    k2.overwrite(k);
    StateDel.overwrite(State); 
    StateDel.plus_mult_eq(k2,TIMESTEP/2);
    break;
  case (3):
    //Third Call
    //Derivatives(StateDel,k3);
    k3.overwrite(k);
    StateDel.overwrite(State);
    StateDel.plus_mult_eq(k3,TIMESTEP);
    break;
  case (4):
    //Fourth Call
    //Derivatives(StateDel,k4);
    k4.overwrite(k);
    //Put it all together
    // phi = k1/6 + k2/3 + k3/3 + k4/6
    phi.mult(k1,1.0/6.0); // phi = k1/6
    phi.plus_mult_eq(k2,1.0/3.0); // phi = phi + k2/3
    phi.plus_mult_eq(k3,1.0/3.0); // phi = phi + k3/3
    phi.plus_mult_eq(k4,1.0/6.0); // phi = phi + k2/3
    //And then State = State + phi*timestep
    State.plus_mult_eq(phi,TIMESTEP);
  }
}

File: libraries/RK4/RK4.py
import numpy as np


class RK4():
    """Literal port of RK4.cpp: a vehicle-agnostic 4-stage RK4 integrator.

    The caller is responsible for computing the derivative vector k
    (matching integrator.k in the C++ side) before each integrate(i) call.
    """

    def __init__(self):
        pass

    def init(self, numvars, dt):
        self.NUMVARS = numvars
        self.TIMESTEP = dt
        self.State = np.zeros(numvars)
        self.StateDel = np.zeros(numvars)
        self.k = np.zeros(numvars)
        self.k1 = np.zeros(numvars)
        self.k2 = np.zeros(numvars)
        self.k3 = np.zeros(numvars)
        self.k4 = np.zeros(numvars)
        self.phi = np.zeros(numvars)

    def set_ICs(self, x0):
        self.State = np.array(x0, dtype=float).copy()
        self.StateDel = np.array(x0, dtype=float).copy()

    def integrate(self, i):
        dt = self.TIMESTEP
        if i == 1:
            self.k1 = self.k.copy()
            self.StateDel = self.State + self.k1 * (dt / 2.0)
        elif i == 2:
            self.k2 = self.k.copy()
            self.StateDel = self.State + self.k2 * (dt / 2.0)
        elif i == 3:
            self.k3 = self.k.copy()
            self.StateDel = self.State + self.k3 * dt
        elif i == 4:
            self.k4 = self.k.copy()
            self.phi = self.k1 / 6.0 + self.k2 / 3.0 + self.k3 / 3.0 + self.k4 / 6.0
            self.State = self.State + self.phi * dt


if __name__ == '__main__':
    ##Integrate dx/dt = 1 (constant derivative) for 1 second, expect x(1.0) ~= 1.0
    rk = RK4()
    rk.init(1, 0.01)
    rk.set_ICs([0.0])
    t = 0.0
    while t < 1.0 - 1e-9:
        for i in (1, 2, 3, 4):
            rk.k[0] = 1.0
            rk.integrate(i)
        t += 0.01
    assert abs(rk.State[0] - 1.0) < 1e-6, rk.State[0]
    print('RK4.py OK: x(1.0) =', rk.State[0])
